### ISSUE TITLE
Numerical sensitivity in GroupNorm backward pass under upstream-gradient scaling (metamorphic testing) in TF2.17.0

### ISSUE THREAD

**Issue by XiaoHuaOvo-tech** [tensorflow/tensorflow#107788]
### Issue type

Performance

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TF2.17.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

3.10.19

### CUDA/cuDNN version

CUDA version: 12.3 cuDNN version: 8

### GPU model and memory

Tesla V100

### Current behavior?

We are using a gradient metamorphic testing framework to validate operator backprop. As a control operator, we implemented GroupNorm using standard TF ops (reshape, tf.nn.moments, sqrt, affine). We test a basic linearity MR:

For a fixed forward graph and upstream gradients G1 and G2 = δ·G1, gradients should satisfy
∂y/∂θ (G2) ≈ δ · ∂y/∂θ (G1).

We quantify violations using a tolerance-normalized metric (atol=1e-5, rtol=1e-3). We observe:

TF 2.17.0: the same tests show much smaller/sparser violations.vR ≈ 287.

Additivity MR (g(G1+G2) == g(G1)+g(G2)) passes exactly.

Permutation equivariance and degeneration consistency also pass.

Questions:

Is this level of “scaling linearity” deviation in GroupNorm gradients expected due to float32 reduction order / non-associativity or backend-specific kernels?

Are there recommended flags or execution modes to improve determinism/stability (e.g., deterministic ops, XLA, oneDNN settings)?

We provide a minimal script and vR outputs if useful.

We do not believe this indicates a correctness bug in GroupNorm.

Rather, we are interested in understanding whether this numerical sensitivity is expected, and whether there are recommended practices or

configuration options to improve stability or determinism in such cases.

### Standalone code to reproduce the issue

```shell
import numpy as np
import tensorflow as tf

# ---------------- vR metric ----------------
def compute_vR(A, B, atol=1e-5, rtol=1e-3):
    A = np.asarray(A).flatten()
    B = np.asarray(B).flatten()
    diff = np.abs(A - B)
    scale = np.maximum(np.abs(A), np.abs(B))
    tol = atol + rtol * scale
    r = diff / (tol + 1e-300)
    return float(np.max(r)), float(np.max(diff))

np.random.seed(0)
tf.random.set_seed(0)

print("TensorFlow version:", tf.__version__)

N, H, W, C = 2, 16, 16, 32
num_groups = 2
eps = 1e-4

x = tf.ones((N, H, W, C), tf.float32) * 10.0
x += tf.random.normal(x.shape, stddev=1e-3)

gamma = tf.ones([C], tf.float32)
beta  = tf.zeros([C], tf.float32)

# ==========================================
# GroupNormalization 
# ==========================================
def groupnorm_backward(x, gamma, beta, G):
    x_v = tf.Variable(x)
    

    gn_layer = tf.keras.layers.GroupNormalization(
        groups=num_groups,
        axis=-1,
        epsilon=eps,
        center=True, 
        scale=True,
        gamma_initializer=tf.constant_initializer(gamma.numpy()),
        beta_initializer=tf.constant_initializer(beta.numpy())
    )

    with tf.GradientTape() as tape:
        tape.watch(x_v)
        y = gn_layer(x_v)
    gX = tape.gradient(y, x_v, output_gradients=G)
    return gX

# ---------- GMR-1 ----------
G1 = tf.ones_like(x)
delta = -3.0
G2 = delta * G1

# 计算梯度
gX1 = groupnorm_backward(x, gamma, beta, G1)
gX2 = groupnorm_backward(x, gamma, beta, G2)

# 计算 vR
vR, max_abs = compute_vR(
    gX2.numpy(),
    (delta * gX1).numpy()
)

print("\n[Native GN-GMR-1: input gradient scaling]")
print("delta =", delta)
print("max |gX(δG) - δ·gX(G)| =", max_abs)
print("vR =", vR)

print("mean |gX(G)| =", np.mean(np.abs(gX1.numpy())))
print("mean |error| =", np.mean(np.abs(gX2.numpy() - delta * gX1.numpy())))
```

### Relevant log output

```shell
[Native GN-GMR-1: input gradient scaling]
delta = -3.0
max |gX(δG) - δ·gX(G)| = 0.003885726910084486
vR = 287.0324401855469
mean |gX(G)| = 0.00057874783
mean |error| = 0.0005262308
```

**Comment by Somanshu-Jha** [comment]
Hi, I would like to work on this issue.
Please assign it to me.

**Comment by nileshau-afk** [comment]
Hi @XiaoHuaOvo-tech , @Somanshu-Jha ,

We truly appreciate your support. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

We look forward to seeing your contributions!
